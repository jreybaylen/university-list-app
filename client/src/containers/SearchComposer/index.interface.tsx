interface SearchComposerProps {
    keyType: string
    keyWord: string
    onSubmit(param: string | object): void
}

export type {
    SearchComposerProps
}