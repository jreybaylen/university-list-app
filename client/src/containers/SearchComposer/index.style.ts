import { StypesProps } from '@interface/style.interface'

const styles: StypesProps = {
    container: {
        padding: 20,
        width: '100%',
        maxWidth: '500px'
    },
    searchForm: {
        display: 'flex'
    }
}

export {
    styles
}