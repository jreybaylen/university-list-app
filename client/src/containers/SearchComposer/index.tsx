import { useState, ChangeEvent } from 'react'

import { SearchComposerProps } from './index.interface'

import { Input, Form } from '@components/index'

function SearchComposer (props: SearchComposerProps): JSX.Element {
    const [ search, setSearch ] = useState<string>('')
    const handleChangeSearch = (event: ChangeEvent<HTMLInputElement>) => {
        setSearch(event.target.value)
    }
    const handleSearchSubmit = () => {
        if (props.onSubmit) {
            props.onSubmit(search)
        }
    }
    const searchComposerElement = (
        <Form onSubmit={ handleSearchSubmit }>
            <Input
                value={ search }
                placeholder="Search"
                onChange={ handleChangeSearch }
            />
        </Form>
    )

    return searchComposerElement
}

export default SearchComposer