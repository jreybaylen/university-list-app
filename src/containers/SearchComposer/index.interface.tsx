import { FormProps } from '@interface/form.interface'

interface SearchComposerProps {
    keyType: string
    keyWord: string
    onSubmit(param: string | object): void
}

export type {
    FormProps,
    SearchComposerProps
}