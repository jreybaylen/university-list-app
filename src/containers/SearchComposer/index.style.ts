import { StypesProps } from '@interface/style.interface'

const styles: StypesProps = {
    container: {
        padding: 20,
        maxWidth: 600,
        width: '100%',
        margin: 'auto'
    },
    searchForm: {
        display: 'flex'
    },
    input: {
        textTransform: 'capitalize'
    }
}

export {
    styles
}