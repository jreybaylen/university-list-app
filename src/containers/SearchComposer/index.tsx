import { useState, useEffect, useRef, ChangeEvent } from 'react'

import { styles } from './index.style'
import { SearchComposerProps } from './index.interface'

import { Input, Form, Picker, Button } from '@components/index'

function SearchComposer (props: SearchComposerProps): JSX.Element {
    const [ type, setType ] = useState<string>('')
    const [ search, setSearch ] = useState<string>('')
    const handleChangeType = (event: ChangeEvent<HTMLSelectElement>) => {
        const { target: { value } } = event

        setType(value)
    }
    const handleChangeWord = (event: ChangeEvent<HTMLInputElement>) => {
        const { target: { value } } = event

        setSearch(value)
    }
    const handleSearchSubmit = () => {
        if (props.onSubmit) {
            props.onSubmit({ search, type })
        }
    }
    const searchComposerInit = useRef(() => {
        const { keyType, keyWord } = props

        setSearch(keyWord)
        setType(keyType.toLocaleLowerCase())
    })

    useEffect(() => {
        searchComposerInit.current()
    }, [])

    const searchComposerElement = (
        <div style={ styles.container }>
            <Form onSubmit={ handleSearchSubmit }>
                <div style={ styles.searchForm }>
                    <Picker
                        value={ type }
                        onChange={ handleChangeType }
                        options={ [ 'Name', 'Country' ] }
                    />
                    <Input
                        value={ search }
                        placeholder="Search"
                        onChange={ handleChangeWord }
                        style={ styles.input }
                    />
                    <Button type="submit" children="Go" />
                </div>
            </Form>
        </div>
    )

    return searchComposerElement
}

export default SearchComposer