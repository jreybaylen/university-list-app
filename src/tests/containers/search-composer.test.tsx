import userEvent from '@testing-library/user-event'
import { render, fireEvent } from '@testing-library/react'

import SearchComposer from '@container/SearchComposer'

const keyType = 'country'
const keyWord = 'Philippines'

describe('<SearchComposer /> Component', () => {
    it('Should render without crashing', () => {
        const handleSubmit = jest.fn()

        render(
            <SearchComposer
                keyType={ keyType }
                keyWord={ keyWord }
                onSubmit={ handleSubmit }
            />
        )
    })

    it('Should render Form, Picker, Input and Button component', () => {
        const handleSubmit = jest.fn()
        const { container } = render(
            <SearchComposer
                keyType={ keyType }
                keyWord={ keyWord }
                onSubmit={ handleSubmit }
            />
        )

        expect(container.querySelector('form')).toBeInTheDocument()
        expect(container.querySelector('select[name="type"]')).toBeInTheDocument()
        expect(container.querySelector('input[name="word"]')).toBeInTheDocument()
        expect(container.querySelector('button[type="submit"]')).toBeInTheDocument()
    })

    it('Should trigger onSubmit once the form has been submitted', () => {
        const handleSubmit = jest.fn()
        const { getByText } = render(
            <SearchComposer
                keyType={ keyType }
                keyWord={ keyWord }
                onSubmit={ handleSubmit }
            />
        )

        expect(handleSubmit).not.toHaveBeenCalled()
        fireEvent.submit(getByText(/go/i))
        expect(handleSubmit).toHaveBeenCalled()
    })

    it('Should change the value of Picker once the picker has been changed', () => {
        const handleSubmit = jest.fn()
        const { container, getByDisplayValue } = render(
            <SearchComposer
                keyType={ keyType }
                keyWord={ keyWord }
                onSubmit={ handleSubmit }
            />
        )
        const optionElement = getByDisplayValue(/name|country/i)
        const selectElement = container.querySelector('select[name="type"]')
        const nameOption = container.getElementsByTagName('option').item(0)
        const countryOption = container.getElementsByTagName('option').item(1)

        expect(optionElement).toBeInTheDocument()
        userEvent.selectOptions(selectElement as HTMLSelectElement, 'name')
        expect((nameOption as HTMLOptionElement).selected).toBe(true)
        userEvent.selectOptions(selectElement as HTMLSelectElement, 'country')
        expect((countryOption as HTMLOptionElement).selected).toBe(true)
        expect((nameOption as HTMLOptionElement).selected).toBe(false)
        expect((selectElement as HTMLSelectElement).value).toStrictEqual('country')
    })
})