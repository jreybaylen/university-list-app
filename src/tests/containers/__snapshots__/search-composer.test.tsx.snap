// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchComposer /> Component Should render without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="padding: 20px; max-width: 600px; width: 100%; margin: auto;"
      >
        <form
          name="search"
        >
          <div
            style="display: flex;"
          >
            <select
              id="picker-component"
              name="type"
              style="outline: 0; cursor: pointer; padding: 10px 15px; border-top-left-radius: 25px; border: 1px solid #ccc; border-bottom-left-radius: 25px; text-transform: capitalize;"
            >
              <option
                value="name"
              >
                Name
              </option>
              <option
                value="country"
              >
                Country
              </option>
            </select>
            <input
              name="word"
              placeholder="Search"
              style="outline: 0; width: 100%; font-weight: normal; padding: 10px 15px; border: 1px solid #ccc; text-transform: capitalize;"
              value="Philippines"
            />
            <button
              style="outline: 0; cursor: pointer; padding: 10px 15px; border-top-right-radius: 25px; border: 1px solid #ccc; border-bottom-right-radius: 25px; text-transform: capitalize;"
              type="submit"
            >
              Go
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="padding: 20px; max-width: 600px; width: 100%; margin: auto;"
    >
      <form
        name="search"
      >
        <div
          style="display: flex;"
        >
          <select
            id="picker-component"
            name="type"
            style="outline: 0; cursor: pointer; padding: 10px 15px; border-top-left-radius: 25px; border: 1px solid #ccc; border-bottom-left-radius: 25px; text-transform: capitalize;"
          >
            <option
              value="name"
            >
              Name
            </option>
            <option
              value="country"
            >
              Country
            </option>
          </select>
          <input
            name="word"
            placeholder="Search"
            style="outline: 0; width: 100%; font-weight: normal; padding: 10px 15px; border: 1px solid #ccc; text-transform: capitalize;"
            value="Philippines"
          />
          <button
            style="outline: 0; cursor: pointer; padding: 10px 15px; border-top-right-radius: 25px; border: 1px solid #ccc; border-bottom-right-radius: 25px; text-transform: capitalize;"
            type="submit"
          >
            Go
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
